Cross-compile Instruction for Linux targeting Windows (#1114)
htop-dev/htop: docs/winnt/cross-build-for-mingw64.md

```
./configure \
	--host=$HOST_PLATFORM \
	--prefix=$MINGW_ROOT \
	--disable-getcap \
	--disable-hard-tabs \
	--disable-home-terminfo \
	--disable-leaks \
	--disable-macros \
	--disable-overwrite \
	--disable-shared
	--disable-termcap \
	--enable-assertions \
	--enable-database \
	--enable-echo   \
	--enable-exp-win32 \
	--enable-ext-funcs \
	--enable-interop \
	--enable-opaque-curses \
	--enable-opaque-form \
	--enable-opaque-menu \
	--enable-opaque-panel \
	--enable-pc-files \
	--enable-sp-funcs \
	--enable-static\
	--enable-term-driver \
	--enable-warnings \
	--enable-widec \
	--with-fallbacks=ms-terminal \
	--with-normal \
	--with-progs \
	--without-ada \
	--without-debug \
	--without-libtool \
	--without-manpages \
	--without-shared \
```
This configuration is based on the recommended ones for compilation to Windows from NCurses developers ( [README.MinGW](https://github.com/mirror/ncurses/blob/master/README.MinGW) with the addition of `--enable-widec` to provide Unicode support. Script `build_ncurses.sh` is provided to configure with these options.

> Standard `make` and `make install` to build the library and install it and its headers into `$MINGW_ROOT`

Reviewer: It seems `make install` in this step tries to unlink an existing global system installation of `ncurses`:

```
$ make install
( cd include && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/include“ wird betreten
mkdir -p /mingw-root/build-root/winnt/include/ncursesw
installing curses.h in /mingw-root/build-root/winnt/include/ncursesw
installing unctrl.h in /mingw-root/build-root/winnt/include/ncursesw
installing ncurses_dll.h in /mingw-root/build-root/winnt/include/ncursesw
installing term.h in /mingw-root/build-root/winnt/include/ncursesw
installing termcap.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./tic.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./term_entry.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./nc_tparm.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./win32_curses.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/include“ wird verlassen
( cd ncurses && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/ncurses“ wird betreten
mkdir -p /mingw-root/build-root/winnt/lib
installing ../lib/libncursesw.a as /mingw-root/build-root/winnt/lib/libncursesw.a
/usr/bin/install -c -m 644 ../lib/libncursesw.a /mingw-root/build-root/winnt/lib/libncursesw.a
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/ncurses“ wird verlassen
( cd progs && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/progs“ wird betreten
mkdir -p /mingw-root/build-root/winnt/bin
/usr/bin/install -c -s tic.exe     /mingw-root/build-root/winnt/bin/`echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s toe.exe     /mingw-root/build-root/winnt/bin/`echo toe.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
linking infotocap.exe to tic.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo infotocap.exe| sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
linking captoinfo.exe to tic.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo captoinfo.exe| sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
/usr/bin/install -c -s infocmp.exe /mingw-root/build-root/winnt/bin/`echo infocmp.exe|   sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s clear.exe   /mingw-root/build-root/winnt/bin/`echo clear.exe|     sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tabs.exe    /mingw-root/build-root/winnt/bin/`echo tabs.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tput.exe    /mingw-root/build-root/winnt/bin/`echo tput.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tset.exe    /mingw-root/build-root/winnt/bin/`echo tset.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
linking reset.exe to tset.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tset.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo reset.exe|     sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/progs“ wird verlassen
( cd panel && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/panel“ wird betreten
installing ../lib/libpanelw.a as /mingw-root/build-root/winnt/lib/libpanelw.a
/usr/bin/install -c -m 644 ../lib/libpanelw.a /mingw-root/build-root/winnt/lib/libpanelw.a
installing ./panel.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/panel“ wird verlassen
( cd menu && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/menu“ wird betreten
installing ../lib/libmenuw.a as /mingw-root/build-root/winnt/lib/libmenuw.a
/usr/bin/install -c -m 644 ../lib/libmenuw.a /mingw-root/build-root/winnt/lib/libmenuw.a
installing ./eti.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./menu.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/menu“ wird verlassen
( cd form && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/form“ wird betreten
installing ../lib/libformw.a as /mingw-root/build-root/winnt/lib/libformw.a
/usr/bin/install -c -m 644 ../lib/libformw.a /mingw-root/build-root/winnt/lib/libformw.a
installing ./form.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/form“ wird verlassen
( cd test && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/test“ wird betreten
make[1]: Für das Ziel „install“ ist nichts zu tun.
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/test“ wird verlassen
( cd misc && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/misc“ wird betreten
mkdir -p /mingw-root/build-root/winnt/share
mkdir -p /mingw-root/build-root/winnt/share/tabset
DESTDIR= \
prefix=/mingw-root/build-root/winnt \
exec_prefix=/mingw-root/build-root/winnt \
bindir=/mingw-root/build-root/winnt/bin \
top_srcdir=.. \
srcdir=. \
datadir=/mingw-root/build-root/winnt/share \
ticdir=/mingw-root/build-root/winnt/share/terminfo \
source=terminfo.tmp \
cross_compiling=yes \
/bin/sh ./run_tic.sh
** Building terminfo database, please wait...
Running /usr/bin/tic to install /mingw-root/build-root/winnt/share/terminfo ...

        You may see messages regarding extended capabilities, e.g., AX.
        These are extended terminal capabilities which are compiled
        using
                tic -x
        If you have ncurses 4.2 applications, you should read the INSTALL
        document, and install the terminfo without the -x option.

ncurses 6.3.20211021
1788 entries written to /mingw-root/build-root/winnt/share/terminfo
** built new /mingw-root/build-root/winnt/share/terminfo
** sym-linked /mingw-root/build-root/winnt/lib/terminfo for compatibility
installing std
installing stdcrt
installing vt100
installing vt300
finished install.data
/usr/bin/install -c ncurses-config /mingw-root/build-root/winnt/bin/ncursesw6-config
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing panelw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/panelw.pc': Permission denied
installing menuw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/menuw.pc': Permission denied
installing formw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/formw.pc': Permission denied
installing ncurses++w.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncurses++w.pc': Permission denied
make[1]: *** [Makefile:150: install] Fehler 1
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/misc“ wird verlassen
make: *** [Makefile:138: install] Fehler 2
```

Author: I've run into this file testing.  It's trying to install the pkg-config files for the mingw64 version to the global system because that's where pkg-config tells it to.  pkg-config should be disabled. I'll take a look at it, could attach your config.log?

Reviewer: [config.log for ncurses](https://github.com/htop-dev/htop/files/9930220/config.log)

In my case it's properly installing the files in the proper platform directory (see start of the `make install` output), but further down it operating on files of the global system. With a prefix set this should not happen.

Author: These are the errors I'm seeing:

```
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing panelw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/panelw.pc': Permission denied
installing menuw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/menuw.pc': Permission denied
installing formw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/formw.pc': Permission denied
installing ncurses++w.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncurses++w.pc': Permission denied
make[1]: *** [Makefile:150: install] Fehler 1
```

This is trying to install the pkg-config files for mingw64, but it's getting the location to put those files from running `pkg-config` instead basing it on the prefix. Getting pkg-config setup properly just seems unnecessary. Looking at your `config.log`  the wrong options are being given to `configure` . It should be disabling pkg-config stuff from being built (and then installed).  e.g., `--enable-pc-files` should be the opposite. 

Cleaning up the ncurses options along with testing that it was building binaries that would run properly on Windows was the last thing I did before my git adventure. I think I might have dropped a commit with the final version of the flags. I'll look at it and fix it.

Author: Was able to take a quick look. The `build_ncurses.sh` script and the configure options in the instructions are out of sync, with the script having the correct ones (from a quick look).

The scripts have been handy when I've been working, but would it be best to remove them?  If only to remove the possibility of them being out of sync with the documentation?

Reviewer: I only looked at the instructions earlier when I tested this.

Given that documentation should usually not contain executables I'd suggest to remove those scripts.

## Configure and Cross-compile htop

This should be everything we need to cross compile `htop` itself.

Clone the `htop` repo and checkout the `windows` branch (currently work on the `winnt` Port is separated into that branch):

```
    git clone https://github.com/htop-dev/htop.git
    cd htop
    git checkout windows
```

To configure `htop` for cross-compiling we just need to tell autoconf we want to target the desired platform, and make sure that it looks to our `$MINGW_ROOT` for ncurses. Script `build_htop.sh` is provided to configure with these options.

```
./autogen.sh

> # Make sure configure can find the cross0built ncurses library and headers

Reviewer: ```suggestion
# Make sure configure can find the cross-built ncurses library and headers
```

export CFLAGS="-I$MINGW_ROOT/include"
export LDFLAGS="-L$MINGW_ROOT/lib"

./configure \
	--host=$HOST_PLATFORM \
	--prefix=$MINGW_ROOT \
	--enable-static \

> 	--enable-unicode

Reviewer: This fails to find `sys/utsname.h`, as that one is listed in `configure.ac` to be required. Instead when targeting platform `winnt` this should be skipped in `configure.ac`.

Workaround:

```diff
diff --git a/configure.ac b/configure.ac
index 53163366..9893f500 100644
--- a/configure.ac
+++ b/configure.ac
@@ -138,7 +138,6 @@ AC_CHECK_HEADERS([ \
     strings.h \
     sys/param.h \
     sys/time.h \
-    sys/utsname.h \
     unistd.h
    ], [], [AC_MSG_ERROR([can not find required generic header files])])
 
@@ -160,6 +159,10 @@ if test "$my_htop_platform" = darwin; then
    AC_CHECK_HEADERS([mach/mach_time.h])
 fi
 
+if test "$my_htop_platform" != unsupported; then
+   AC_CHECK_HEADERS([sys/utsname.h], [], [AC_MSG_ERROR([can not find required generic header files])])
+fi
+
 # ----------------------------------------------------------------------
 
 
```

Ideally this should be handled as a separate `winnt` platform, but this is upcoming work.

Reviewer 2: > Workaround:

FYI gnulib has a stub for this, See: https://www.gnu.org/software/gnulib/manual/html_node/sys_002futsname_002eh.html and https://github.com/coreutils/gnulib/blob/f36daeac96c30aa5ed42ffa9ccd8cd4366b08b13/lib/uname.c#L56

Author: > > Workaround:
> 
> FYI gnulib has a stub header and implementation for this on MinGW/MSVC, See: https://www.gnu.org/software/gnulib/manual/html_node/sys_002futsname_002eh.html and https://github.com/coreutils/gnulib/blob/f36daeac96c30aa5ed42ffa9ccd8cd4366b08b13/lib/uname.c#L56

The `winnt` build shouldn't be using `utsname.h` (it's functionality isn't available on Windows). We want its contents to be missing so that builds that include break ASAP.

```

> Then the standard `make` and `make install` to build and put the files into the $MINGW_ROOT tree.

Reviewer: This fails to compile in several places due to some headers implicitly used:
- `langinfo.h`
- `pwd.h`
- `sys/resource.h`
- `sys/select.h`
- `sys/wait.h`

Furthermore some defines/symbols are not present:
- `AT_FDCWD`
- `AT_SYMLINK_NOFOLLOW`
- `lstat`
- `readlink`

Overall this list seems manageable.

# Cross-compiling htop for Windows from Linux

> An intermediate build using Cygwin/MSYS is being worked on by @ookiineko as of 2023-09-28 (see #1097)

Reviewer 2: while you can use them as a starting point for the native port, I think Cygwin/Msys ports are NOT intermediate builds, they should be considered as standalone unix-like platforms. See: https://github.com/htop-dev/htop/issues/1097#issuecomment-1740194685

Author: I think their existence should be mentioned in this document.  Not sure on the best wording.

I don't agree with your view of the platforms or usefulness them beyond a native port, but that have nothing to do with this PR.

## Setting up the Cross-Build Environment

> Going through specifics, we going assume a base install of Ubuntu 22.04 for `x86_64`. The same version used by GitHub's ubuntu-latest runner. It has (almost) everything needed in its packaging system. While package names may differ,  most major distributions will have same tools available.

Reviewer: ```suggestion
Going through specifics, we are going to assume a base install of Ubuntu 22.04 for `x86_64`; the same version also used by GitHub's `ubuntu-latest` runner. It has (almost) everything needed in its packaging system. While package names may differ, most major distributions will have these tools available.
```

    git checkout windows
```

To configure `htop` for cross-compiling we just need to tell autoconf we want to target the desired platform, and make sure that it looks to our `$MINGW_ROOT` for NCurses.

```
./autogen.sh

> # Make sure configure can find the cross built NCurses library and headers

Reviewer: ```suggestion
# Make sure configure can find the cross-built NCurses library and headers
```
